sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CustomShuffleReaderExec.scala

```diff
@@ -179,12 +179,12 @@ case class CustomShuffleReaderExec private(
   }
 
   private lazy val shuffleRDD: RDD[_] = {
-    sendDriverMetrics()
```

Comment:
So, this is a regression from SPARK-31253 (https://github.com/apache/spark/commit/d136b7248ecc7303f86da8f68e0625d9d0725a50)? Do you think we can have a test case for `NoSuchElementException`?
```scala
- val executionId = sparkContext.getLocalProperty(SQLExecution.EXECUTION_ID_KEY)
- SQLMetrics.postDriverMetricUpdates(sparkContext, executionId, metrics.values.toSeq)
+ sendDriverMetrics()
```

Reply:
I looked into writing a test. I would like to do something like this in `AdaptiveQueryExecSuite`:

```scala
  test("operating on canonicalized plan") {
    withSQLConf(SQLConf.ADAPTIVE_EXECUTION_ENABLED.key -> "true") {
      val (_, adaptivePlan) = runAdaptiveAndVerifyResult(
        "SELECT key FROM testData GROUP BY key")
      val readers = collect(adaptivePlan) {
        case r: CustomShuffleReaderExec => r
      }
      assert(readers.length == 1)
      val reader = readers.head
      val ex = intercept[IllegalStateException] {
        val c = reader.canonicalized.asInstanceOf[CustomShuffleReaderExec]
        c.doExecute()
      }
      assert(ex.getMessage === "operating on canonicalized plan")
    }
  }
```

The problem is that `doExecute` is protected, and I can't test by calling `execute` because it has separate checks for canonicalized plans.

What would be the recommended approach here?

Reply:
I went ahead and pushed a working test. I used Java reflection to call the protected method.